lib/fine-types/src/Language/FineTypes/Export/OpenAPI/Schema.hs

```diff
@@ -0,0 +1,278 @@
+    , Referenced (Inline, Ref)
+    , Schema
+        ( _schemaAdditionalProperties
+        , _schemaAllOf
+        , _schemaFormat
+        , _schemaItems
+        , _schemaMinimum
+        , _schemaOneOf
+        , _schemaProperties
+        , _schemaRequired
+        , _schemaTitle
+        , _schemaType
+        )
+    )
+import Language.FineTypes.Module
+    ( Declarations
+    , Module (..)
+    , resolveVars
+    )
+import Language.FineTypes.Typ
+    ( ConstructorName
+    , FieldName
+    , OpOne (..)
+    , OpTwo (..)
+    , Typ (..)
+    , TypConst (..)
+    , everything
+    , everywhere
+    )
+
+import qualified Data.HashMap.Strict.InsOrd as MH
+import qualified Data.Map as Map
+import qualified Data.Text as T
+
+{-----------------------------------------------------------------------------
+    OpenAPI
+------------------------------------------------------------------------------}
+
+-- | Export
+--
+-- Assumes that the argument satisfies 'supportsJSON'.
+schemaFromModule :: Module -> OpenApi
+schemaFromModule m = mempty{_openApiComponents = ds, _openApiInfo = info'}
+  where
+    info' =
+        mempty
+            { _infoTitle = T.pack $ moduleName m
+            }
+    ds = mapDeclarations $ moduleDeclarations m
+
+mapDeclarations :: Declarations -> Components
+mapDeclarations ds = mkComponents $ do
+    (typName, typ) <- Map.toList ds
+    pure (T.pack typName, schemaFromTyp typ)
+  where
+    mkComponents xs = mempty{_componentsSchemas = MH.fromList xs}
+
+-- | Test whether a 'Module' only uses types supported by JSON.
+--
+-- JSON does not support finite maps such as @↦@, @↦0@, @→∗@.
+supportsJSON :: Module -> Bool
+supportsJSON =
+    and . Map.map isSupportedTyp . moduleDeclarations
+  where
+    isSupportedTyp = everything (&&) isSupported
+    isSupported (Two fun _ _) =
+        fun `notElem` [PartialFunction, FiniteSupport]
+    isSupported (Constrained _ _) =
+        False
+    isSupported _ = True
+
+-- | Convert 'Typ' definitions to JSON.
+--
+-- The result satisfies 'supportsJSON'.
+--
+-- Note: We don't recommend that you use this function,
+-- because it does a lot of conversions under the hood.
+-- Instead, if you want to export a 'Typ' to JSON,
+-- we recommend that you explicitly define a second 'Typ'
+-- which is apparently compatible with JSON,
+-- and show that the first 'Typ' can be embedded into the second 'Typ'.
+convertToJSON :: Declarations -> Declarations
+convertToJSON declarations = Map.map (jsonify declarations) declarations
+
+{-----------------------------------------------------------------------------
+    Convert Typ to JSON schema
+------------------------------------------------------------------------------}
+
+schemaFromTyp :: Typ -> Schema
+schemaFromTyp = go
+  where
+    go Abstract =
+        mempty
+            { _schemaType = Just OpenApiObject
+            }
+    go (Var name') =
+        mempty
+            { _schemaAllOf = Just [Ref $ Reference $ T.pack name']
```

Comment:
I checked that in conjunction with `mapDeclarations`, using the name `Reference $ T.pack name'` is sufficient to produce a valid schema reference — the path prefix `#/components/schemas/` will be inserted automatically by the `openapi3` library when exporting to JSON via `ToJSON`, see

https://hackage.haskell.org/package/openapi3-3.2.3/docs/src/Data.OpenApi.Internal.html#line-1430

👍

Reply:
I just had a proof of it in https://github.com/cardano-foundation/fine-types/pull/24 :-)